### PR TITLE
PyTorch 2.11 X86 Migration - SageMaker

### PR DESCRIPTION
**Pull request by Eren-Jeager123**
### What

First phase of migrating SageMaker tests from V1 to V2. Adds the SageMaker image target and 2 basic P0 integration tests. The full PR workflow (build → sanity → security → unit → SM tests) takes ~45 min.

### Dockerfile

- Restructured stages: `runtime-base` → `runtime` (EC2/ECS) and `sagemaker` (SM Training)
- Removed `runtime-eks` stage
- Added `sagemaker` stage: SM PySdk v3, sagemaker-pytorch-training, smclarify, data science packages, hostname workaround
- SageMaker packages defined in `pyproject.toml` `[project.optional-dependencies.sagemaker]`, installed via `uv sync --extra sagemaker`
- Pinned `mlflow>=3.9.0` and `sagemaker>=3.4.0` to resolve 42 CRITICAL/HIGH CVEs from ECR scan
- Moved SageMaker BYOC paths (`/opt/ml/*`) from runtime to sagemaker stage
- Added `start_cuda_compat.sh` to runtime-base (shared by EC2 + SM)
- Added `fastai` to `pyproject.toml` (matching DLC V1)
- Upgraded `pillow` 12.1.1 → 12.2.0 (CVE fix)

### SageMaker test coverage

V1 had 36 SM tests. We profiled them into 7 functionalities at 3 priority levels:

| Priority | Functionality | Status |
|----------|--------------|--------|
| P0 | F1: Distributed training (nccl) | ✅ `test_mnist_distributed_gpu` |
| P0 | F2: torch.distributed primitives | ✅ `test_dist_operations_gpu` |
| P1 | F3: EFA connectivity | Follow-up |
| P1 | F4: GDRCopy (GPU Direct RDMA) | Follow-up |
| P1 | F5: fastai training | Follow-up |
| P2 | F6: torch.compile/inductor on SM | Covered by EC2 tests |
| P2 | F7: Heterogeneous Clusters API | Low value |

Tests use SM PySdk v3 `ModelTrainer` API (V1 used deprecated `PyTorch` Estimator). Both tests pass `FI_EFA_FORK_SAFE=1` via `environment` to prevent Libfabric EFA provider SIGABRT on fork. MNIST training data stored in `s3://dlc-cicd-models/pytorch/sagemaker-test-data/`.

```
CI runner (default-runner, no GPU)
  └── test_sm_training.py (orchestrator)
        ├── test_mnist_distributed_gpu()
        │     └── ModelTrainer.train()
        │           → SageMaker provisions 2x ml.g4dn.12xlarge
        │           → Pulls Docker image from CI ECR
        │           → Runs mnist.py inside container on both nodes
        │           → mnist.py: init distributed, train MNIST with DDP + nccl
        │           → Job completes → test passes
        │
        └── test_dist_operations_gpu()
              └── ModelTrainer.train()
                    → Same flow, runs distributed_operations.py
                    → Tests broadcast, all_reduce, reduce, all_gather with assertions
                    → Job completes → test passes
```

### Workflow

- New `pr-pytorch-sagemaker.yml`: build `--target sagemaker` → sanity (incl. SM label checks) → security → unit → SM integration tests on `ml.g4dn.12xlarge`
- Added `customer-type: sagemaker` and `framework-version: "2.11.0"` to sanity/security test inputs
- Added SageMaker DLC labels via `--label` flags in build step
- Removed `runtime-eks` build, EKS unit tests, and `test_eks.py` from EC2 workflow

### Infrastructure changes (outside this repo)

- Increased `ml.g4dn.12xlarge` SageMaker service quota in cicd account
- Uploaded MNIST test data to `s3://dlc-cicd-models/pytorch/sagemaker-test-data/`

### Follow-up

- Release workflows (EC2 + SageMaker) — not yet implemented, cannot release currently
- Add CPU Dockerfile (`.cpu` / `.gpu` split like V1) and CPU SM tests (gloo backend)
- Add EFA and GDRCopy SM tests on p4d instances (P1 — validates NCCL + EFA + GDRCopy stack under SM entrypoint)
- Add fastai SM test (P1)
- Evaluate remaining V1 SM tests for migration (inductor, torch_distributed throughput)

### CODE REVIEW

File: docker/pytorch/Dockerfile
Comment (junpuf) [inline docker/pytorch/Dockerfile]: You can remove the "/ Batch / SageMaker" from this

Comment (junpuf) [inline docker/pytorch/Dockerfile]: You can move this line to SageMaker stage

File: docker/pytorch/Dockerfile [hunk 1]
```diff
@@ -210,6 +211,43 @@ ENTRYPOINT ["/usr/local/bin/entrypoint.sh"]
 CMD ["bash"]
 
 
+# ── Stage: sagemaker (SageMaker Training) ────────────────────────────────────
+FROM runtime AS sagemaker
```
Comment (junpuf) [inline docker/pytorch/Dockerfile]: If you are creating a output stage for sagemaker, then you need to make a common stage instead of extending the runtime stage. Setups like "# Telemetry", "# Security patch ", "# OSS compliance", "# Entrypoint" and "# PATH and LD_LIBRARY_PATH" should only exist at the output stage.

You can make the staging like: 
```
runtime-base     -> runtime
                  -> runtime-sagemaker
```

File: docker/pytorch/Dockerfile [hunk 2]
```diff
@@ -210,6 +211,43 @@ ENTRYPOINT ["/usr/local/bin/entrypoint.sh"]
...
+RUN chmod +x /usr/local/bin/start_with_right_hostname.sh
+
+# Security patch
+RUN dnf upgrade -y --security --releasever latest && dnf clean all
+
+ENTRYPOINT ["bash", "-m", "start_with_right_hostname.sh"]
+CMD ["/bin/bash"]
+
+
 # ── Stage: runtime-eks ───────────────────────────────────────────────────────
```
Comment (junpuf) [inline docker/pytorch/Dockerfile]: You can remove the runtime-eks stage

File: docker/pytorch/Dockerfile [hunk 3]
```diff
@@ -210,6 +211,43 @@ ENTRYPOINT ["/usr/local/bin/entrypoint.sh"]
 CMD ["bash"]
 
 
+# ── Stage: sagemaker (SageMaker Training) ────────────────────────────────────
+FROM runtime AS sagemaker
+
+ENV SAGEMAKER_TRAINING_MODULE=sagemaker_pytorch_container.training:main
+
+# SageMaker SDK and training packages
```
Comment (junpuf) [inline docker/pytorch/Dockerfile]: Push these (pip installs) into the pyproject.toml, we don't want any of the pip install show up in the Dockerfile. Keep Sagemaker setups like scripts/pytorch/changehostname.c in the sagemaker stage.

File: test/pytorch/sagemaker/resources/data/training/MNIST/processed/training.pt
Comment (junpuf) [inline test/pytorch/sagemaker/resources/data/training/MNIST/processed/training.pt:1]: These are too large to be saved in the repo. Move them to cicd model s3 bucket

File: test/pytorch/sagemaker/resources/data/training/MNIST/processed/test.pt
Comment (junpuf) [inline test/pytorch/sagemaker/resources/data/training/MNIST/processed/test.pt:1]: These are too large to be saved in the repo. Move them to cicd model s3 bucket

File: test/pytorch/sagemaker/resources/data/training/model
Comment (junpuf) [inline test/pytorch/sagemaker/resources/data/training/model:1]: These are too large to be saved in the repo. Move them to cicd model s3 bucket

File: test/pytorch/integration/sagemaker/resources/distributed_operations.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
+# Copyright 2018-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (junpuf) [inline test/pytorch/integration/sagemaker/resources/distributed_operations.py:1]: sagemaker test directory should be placed inside of `integration` direcotry

File: test/pytorch/sagemaker/profile-pytorch-sagemaker-tests.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
+# PyTorch SageMaker Test Profiling
```
Comment (junpuf) [inline test/pytorch/sagemaker/profile-pytorch-sagemaker-tests.md]: remove development file

Comment (Eren-Jeager123) [reply test/pytorch/sagemaker/profile-pytorch-sagemaker-tests.md]: Removed

**Review by junpuf — APPROVED** [review]
